File: 0023-merge-k-sorted-lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();//min heap
        //(a,b)->Integer.compare(b,a) max heap
        for(int i=0;i<lists.length;i++){
            ListNode cur=lists[i];
            while(cur!=null){
                heap.add(cur.val);
                cur=cur.next;
            }
        }

        if(heap.isEmpty()){
            return null;
        }
        
        ListNode head=null;
        ListNode cur=null;
        while(!heap.isEmpty()){
            if(head == null){
                head= new ListNode(heap.poll());
                cur=head;
            }else{
                cur.next = new ListNode(heap.poll());
                cur=cur.next;
            }
        }
        return head;
    }
}

File: 0023-merge-k-sorted-lists/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* merge(struct ListNode* temp1,struct ListNode* temp2){
    struct ListNode* dummy=(struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode* cur=dummy;
    while(temp1!=NULL && temp2!=NULL){
        if(temp1->val<temp2->val){
            cur->next=temp1;
            temp1=temp1->next;
        }else{
            cur->next=temp2;
            temp2=temp2->next;
        }
        cur=cur->next;
    }
    if(temp1==NULL){
        cur->next=temp2;
    }else{
        cur->next=temp1;
    }
    return dummy->next;
}
struct ListNode* mergeKLists(struct ListNode** lists, int n) {
    struct ListNode* first=NULL;
    for(int i=0;i<n;i++){
        first=merge(first,lists[i]);
    }
    return first;
}
